keps/sig-network/20190415-Autopath API for clusterDNS.md

```diff
@@ -0,0 +1,104 @@
+authors:
+  - "@prameshj"
+owning-sig: sig-network
+participating-sigs:
+  - sig-network
+reviewers:
+  - "@thockin"
+  - "@bowei"
+  - "@johnbelamaric"
+approvers:
+  - "@thockin"
+  - "@bowei"
+  - "@johnbelamaric"
+creation-date: 2019-04-15
+last-updated: 2019-10-28
+status: provisional
+---
+# DNS Autopath in PodSpec
+
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Implementation History](#implementation-history)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+This proposal aims to minimize the number of parallel DNS queries generated by a pod by moving searchpath expansion logic to the DNS server side. This introduces a new dnsPolicy that can be configured on a per-pod basis. The metadata required to complete the searchpath expansion will be sent to the DNS Server via an EDNS0 option. This will be inserted into the request by the Nodelocal DNSCache.
+
+## Motivation
+DNS Search Path expansion on pods using ClusterFirst DNS mode can lead to DNS latency issues and race conditions due to several parallel (The queries are sent in parallel in musl) dns queries from the same pod. Even in pods using glibc which sends these requests serially, the reduced load on client resolver and reduction in client latency is a big motivation to move this logic to the server-side.
+The search path currently includes: 
+
+1. "$NS.svc.$SUFFIX"
+2. "svc.$SUFFIX"
+3. "$SUFFIX"
+4. Host level suffixes, which might be 2 or 3 in number.
+
+Where $NS stands for the namespace that the pod belongs to, $SUFFIX is the Kubernetes cluster suffix.
+
+These search paths are set to make sure: 
+
+ 1. Pods can discover Services in the same namespace using just the service name.
+ 2. Pods can discover Services across namespaces using shorthand of the form "$SVCNAME.$NSNAME"
+ 3. Pods can discover resources within the same cluster.
+
+These searchpaths are included in pods' /etc/resolv.conf by kubelet and are enforced by setting ndots to 5. This means any hostname lookups with fewer than 5 dots will be expanded using all the search paths listed.
+
+When pod issues a query to lookup hostname "service123", it is expanded to 6 queries - one for the original hostname and one with each of the searchpaths appended. Some resolvers issue both A and AAAA queries, so this can be a total of 12 or more queries for every single DNS lookup. When these queries are issued in parallel, they end up at the node with the same source tuple and need to be DNAT'ed increasing the chance of a [netfilter race condition](https://www.weave.works/blog/racy-conntrack-and-dns-lookup-timeouts).
+Even if one of the several queries fails due, the DNS lookup on the client side will fail after a 5s timeout.
```

Comment:
So much time has passed, and now I feel like I see another couple of tweaks.  Not sure if it's correct though.  Is it really true that if any of the queries fails, but some others succeed that the client sees a timeout?
```suggestion
If one of the several queries fails, the DNS lookup on the client side will fail after a 5s timeout.
```

Reply:
I must have missed this during rebase, thought I fixed this. If even one of the expanded queries timeout, the original query issued by the client will timeout. I have seen this in packet captures, not verified by dropping DNS requests though.

Reply:
Thanks for clarifying/fixing.